docs/design/roadmap.md

```diff
@@ -0,0 +1,168 @@
+  * [Refactor common code into `ksonnet/client`](#refactor-common-code-into-ksonnetclient)
+
+
+## Support third-party registries (beyond Github protocol)
+
+*Planned for 0.9.0* ([#232](https://github.com/ksonnet/ksonnet/issues/232))
+
+Currently ksonnet allows you to `ks generate` prototypes from three sources:
+* **[Standard prototypes]** From ksonnet code itself (see [`systemPrototypes.go`](https://github.com/ksonnet/ksonnet/blob/master/prototype/systemPrototypes.go))
+* **[Standard prototypes]** From the [`incubator`](https://github.com/ksonnet/parts/tree/master/incubator) registry in the `ksonnet/parts` Github repo
+* **[Custom prototypes]** From Github directories that conform to the registry spec (see [docs](https://ksonnet.io/docs/cli-reference#ks-registryks-registry-add) for `ks registry add`)
+
+The last option (`ks registry add`) allows users to write and share their own prototypes. *However, this feature is currently limited to Github.* Users may want more flexibility, e.g. a local-filesystem-based registry for development purposes.
+
+**`ks registry add` functionality should be expanded to support other common protocols (local filesystem, S3, etc).** The only requirement thereafter will be having the proper YAML metadata files (e.g. `registry.yaml` in the root).
+
+## Easily generate third-party registries
+
+*Planned for 0.9.0* ([ #234](https://github.com/ksonnet/ksonnet/issues/234))
+
+Users can create custom, third-party registries, but the `ks registry add` command will only recognize them if they adhere to the registry spec. In other words, they need to have the appropriate YAML files and directory structure.
+
+**A new command (e.g. `ks registry create`) which autogenerates registry scaffolding files, may make this easier and less error prone.**
+
+## Lazily download and update dependencies (`ks install`)
+
+*Planned for 0.9.0* ([#217](https://github.com/ksonnet/ksonnet/issues/217), [#237](https://github.com/ksonnet/ksonnet/issues/237))
+
+Currently, `ks init` fails when the user doesn't have an internet connection, because it relies on two external dependencies:
+* The `k.libsonnet` for the `default` environment
+* The `incubator` registry
+
+To avoid this issue, **ksonnet needs to lazily initialize environments and registries**, rather than trying to set these up right away.
+
+**This plays into the idea of a `ks install` command, which will cut down on the number of version-controlled files for a given ksonnet app**. ksonnet dependencies will be `.gitignore`-ed. *Only their metadata* (e.g. versions) will be tracked.
+
+After a ksonnet app is downloaded (e.g. via `git clone`), `ks install` can be run to download and "flesh out" the necessary environments, registries, and packages. This is similar to `npm install`.
+
+**An associated command (e.g. `ks update`) should allow users to download the latest versions of a dependency.**
+
+## Provide access to "ambient parameters"
```

Comment:
The meaning of ambient parameters isn't immediately obvious. Is there a better way to phrase this?